### DISCUSSION TITLE
Support optional ingredients

### DISCUSSION THREAD

**Discussion by schwarz** [cooklang/spec#50] [Ideas]
Sometimes a specific ingredient is nice to have, but doesn't make or break the dish unlike essential ingredients.

**Comment by luizribeiro** [comment]
This could potentially be helpful for documenting possible ingredient replacements as well? I just found myself writing something like `Add @parsley or @mint`. Though you don't need both, just one of them would be fine.

Maybe `Add @parsley or @?mint` could be a good way of writing mint as optional? So a renderer would not render mint as part of the list of ingredients, but would highlight it as an ingredient in the recipe instructions?

**Comment by dubadub** [comment]
> This could potentially be helpful for documenting possible ingredient replacements as well?
I think replacements a bit different. The ingredient can be both optional and have replacements or only optional or required but have replacements.

For replacements I thought to use pipe `Add @parsley | @mint`. Pipe is already used to show alternatives in aisle config and in servings, so for a user meaning should be familiar or guessable. As there should be a way to set different amounts for each replacements we can't use just `Add @parsley | mint` approach.

I like your syntax for optional ingredients `Add @parsley and @?mint`, it communicates meaning well even if someone isn't familiar with the language. I can't think about a better option.

**Comment by chrbala** [comment]
There's a broader use case for more generic variants in recipes. The scope can get quite large if you let it. For example:
* If you have a food processor, use it to chop, otherwise chop by hand
* If you want it spicy, add peppers

In general I think there need to be reasons for why something should or should not be done. Otherwise, people won't know whether they should be doing it or not. Formalized variants could also help determine the requirements ahead of time. For example, a recipe interface could ask you upfront if you want the spicy version or not. Then, it could render the recipe differently depending on the answer.
